source/slang/slang-options.cpp

```diff
@@ -803,6 +805,25 @@ struct OptionsParser
                 {
                     requestImpl->getBackEndReq()->useUnknownImageFormatAsDefault = true;
                 }
+                else if (argStr == "-file-system")
+                {
+                    String name;
+                    SLANG_RETURN_ON_FAIL(tryReadCommandLineArgument(sink, arg, &argCursor, argEnd, name));
+
+                    if (name == "default")
+                    {
+                        spSetFileSystem(compileRequest, nullptr);
+                    }
+                    else if (name == "load-file")
+                    {
+                        spSetFileSystem(compileRequest, OSFileSystem::getSingleton());
```

Comment:
To make sure I understand. The default behavior *used* to be equivalent to using `OSFileSystem`, but now is equivalent to using `OSFileSystemExt`. This is adding a command-line option so that `-file-system load-file` will switch to `OSFileSystem` explicitly, with the old behavior of `OSFileSystem`.

Is this `load-file` case then equivalent to the old behavior? It seems like the answer is going to be "no" because the way that we do/don't wrap things seems like it would be different. I feel like I might benefit from a diagram that breaks down the cases, and what happens in each of them. AFAIU, we have:

* Command-line compiles
  * The default behavior (which is the same as `-file-system default`, right?)
  * And now command-line compile with `-file-sytem load-file`
* API-based compiles
  * No custom file system registered
  * Custom `ISlangFileSystem` registered
  * Custom `ISlangFileSystemExt` registered

So that's five different scenarios I'm aware of (ignoring platform differences in how `OSFileSystem(Ext)` works). From a user perspective, there are two big assumptions I'd be likely to make:

1. That the default `slangc` behavior and the default API behavior are 100% the same
2. That the behavior when specifying an `ISlangFileSystem` instead of an `ISlangFileSystemExt` is 100% equivalent to supplying an `ISlangFileSystemExt` that implements the additonal methods with some known/documented behavior.

Are those reasonable assumptions? I know this isn't the point of the current PR, but opening up a command-line option like this in principle adds more cases, so I want to make sure I understand them all.

Reply:
Also: the cat is already out of the bag, but should we have some convention for marking command-line options that are really just there for testing (e.g., the pass-through stuff, which shouldn't even be exposed via the public API if I'm honest...)?

Reply:
Um. No default behavior has changed. I just added -file-system so I could test out the 'emulation' in tests just too get more code coverage. 

1) Yes
2) No

On 2 -  just so it's clear everything 'slang' works with ISlangFileSystemExt. 

So there are actually 3 scenarios really (the 4th is arguably just for testing)

1) Default: Is the combination of a CacheFileSystem/OSFileSystemExt. Why? Caching improves perf,OSFileSystemExt has the most idea 'what's going on' (like canonicalPaths)
2) User sets ISlangFileSystem. Slang can't deal with that because it only uses ISlangFileSystemExt internally, so the users ISlangFileSystem is wrapped in a CacheFileSystem to improve perf through caching and more importantly emulate all the other methods 
3) User set ISlangFileSystemExt - well that's great. We just use that - we are *assuming* that they know what they are doing and can cache if they want.
4) If file-system load-file we will have a CacheFileSystem on a OSFileSystem

Having a convention for 'test' options seems sensible.

Reply:
Okay. Your option (4) is just for testing, and is really an emulation of (2) in your list.

I think the I was getting at with my second question is that if I as a user wrote:

```
class MyFileSystem : ISlangFileSystem
{ ... }
```

so that I was in case (2) and then I wanted to change to:

```
class MyFileSystem : ISlangFileSystemExt
{ ... /* Ext methods goes here */ }
```

I'm in case (3), and I'm wonder if there is there anything I could (as a user) write for `/* Ext methods go here */` that would give me something (semantically) equivalent to the case (2) behavior I had before.

Similarly, it would be nice to confirm that *if* a user wrote their own `ISlangFileSystemExt` equivalent to our `OSFileSystemExt` that they'd get equivalent behavior to case (1). That ensures that the default behavior really is just "plugging in" a default implementation that the user could have written on their own.

(I'm writing all of this while also assuming that cases (1) and (2) cover 99% of users and case (3) isn't likely to come up)

Reply:
A user could implement the Ext methods and write it such that it equivalent to 2. It is seen as a perhaps uncommon, but useful user feature that they can based on app specific knowledge of what is actually backing the files (zip files, ram, the internet etc). 
That I thought the filesystem as in 'core', such that a user could 'easily' have access to such implementations, but when I look it's actually in slang. It could be in core fairly easily and perhaps should be. 
Our test of 3 is that is is the main 'primary' interface that's used for everything in slang internally. The implementation we test with is 'CacheFileSystem'. We could actually test with OSFileSystem, and for yuks it would easy to add that as a test case for more coverage.